File: sdk/go/nanolink/metrics.go
package nanolink

// Metrics represents system metrics from an agent
type Metrics struct {
	Timestamp    int64            `json:"timestamp"`
	Hostname     string           `json:"hostname"`
	CPU          *CPUMetrics      `json:"cpu,omitempty"`
	Memory       *MemoryMetrics   `json:"memory,omitempty"`
	Disks        []DiskMetrics    `json:"disks,omitempty"`
	Networks     []NetworkMetrics `json:"networks,omitempty"`
	GPUs         []GPUMetrics     `json:"gpus,omitempty"`
	NPUs         []NPUMetrics     `json:"npus,omitempty"`
	UserSessions []UserSession    `json:"userSessions,omitempty"`
	SystemInfo   *SystemInfo      `json:"systemInfo,omitempty"`
	LoadAverage  []float64        `json:"loadAverage,omitempty"`
}

// CPUMetrics represents CPU metrics
type CPUMetrics struct {
	UsagePercent float64   `json:"usagePercent"`
	CoreCount    int       `json:"coreCount"`
	PerCoreUsage []float64 `json:"perCoreUsage,omitempty"`
	Model        string    `json:"model,omitempty"`
	Vendor       string    `json:"vendor,omitempty"`
	FrequencyMHz uint64    `json:"frequencyMhz,omitempty"`
	Temperature  float64   `json:"temperature,omitempty"`
}

// MemoryMetrics represents memory metrics
type MemoryMetrics struct {
	Total     uint64 `json:"total"`
	Used      uint64 `json:"used"`
	Available uint64 `json:"available"`
	SwapTotal uint64 `json:"swapTotal"`
	SwapUsed  uint64 `json:"swapUsed"`
}

// UsagePercent returns memory usage percentage
func (m *MemoryMetrics) UsagePercent() float64 {
	if m.Total == 0 {
		return 0
	}
	return float64(m.Used) / float64(m.Total) * 100
}

// DiskMetrics represents disk metrics
type DiskMetrics struct {
	MountPoint       string  `json:"mountPoint"`
	Device           string  `json:"device"`
	FSType           string  `json:"fsType"`
	Total            uint64  `json:"total"`
	Used             uint64  `json:"used"`
	Available        uint64  `json:"available"`
	ReadBytesPerSec  uint64  `json:"readBytesPerSec"`
	WriteBytesPerSec uint64  `json:"writeBytesPerSec"`
	Model            string  `json:"model,omitempty"`
	DiskType         string  `json:"diskType,omitempty"`
	Temperature      float64 `json:"temperature,omitempty"`
}

// UsagePercent returns disk usage percentage
func (d *DiskMetrics) UsagePercent() float64 {
	if d.Total == 0 {
		return 0
	}
	return float64(d.Used) / float64(d.Total) * 100
}

// NetworkMetrics represents network interface metrics
type NetworkMetrics struct {
	Interface       string   `json:"interface"`
	RxBytesPerSec   uint64   `json:"rxBytesPerSec"`
	TxBytesPerSec   uint64   `json:"txBytesPerSec"`
	RxPacketsPerSec uint64   `json:"rxPacketsPerSec"`
	TxPacketsPerSec uint64   `json:"txPacketsPerSec"`
	IsUp            bool     `json:"isUp"`
	MacAddress      string   `json:"macAddress,omitempty"`
	IPAddresses     []string `json:"ipAddresses,omitempty"`
	SpeedMbps       uint64   `json:"speedMbps,omitempty"`
}

// GPUMetrics represents GPU metrics
type GPUMetrics struct {
	Index           uint32  `json:"index"`
	Name            string  `json:"name"`
	Vendor          string  `json:"vendor"` // NVIDIA, AMD, Intel
	UsagePercent    float64 `json:"usagePercent"`
	MemoryTotal     uint64  `json:"memoryTotal"`
	MemoryUsed      uint64  `json:"memoryUsed"`
	Temperature     float64 `json:"temperature"`
	FanSpeedPercent uint32  `json:"fanSpeedPercent"`
	PowerWatts      uint32  `json:"powerWatts"`
	ClockCoreMHz    uint64  `json:"clockCoreMhz"`
	ClockMemoryMHz  uint64  `json:"clockMemoryMhz"`
	DriverVersion   string  `json:"driverVersion"`
	EncoderUsage    float64 `json:"encoderUsage"`
	DecoderUsage    float64 `json:"decoderUsage"`
}

// NPUMetrics represents NPU/AI accelerator metrics
type NPUMetrics struct {
	Index         uint32  `json:"index"`
	Name          string  `json:"name"`
	Vendor        string  `json:"vendor"` // Intel, Huawei, Qualcomm
	UsagePercent  float64 `json:"usagePercent"`
	MemoryTotal   uint64  `json:"memoryTotal"`
	MemoryUsed    uint64  `json:"memoryUsed"`
	Temperature   float64 `json:"temperature"`
	PowerWatts    uint32  `json:"powerWatts"`
	DriverVersion string  `json:"driverVersion"`
}

// UserSession represents a logged-in user session
type UserSession struct {
	Username    string `json:"username"`
	TTY         string `json:"tty"`
	LoginTime   uint64 `json:"loginTime"`
	RemoteHost  string `json:"remoteHost,omitempty"`
	IdleSeconds uint64 `json:"idleSeconds"`
	SessionType string `json:"sessionType"` // local, ssh, rdp, console
}

// SystemInfo represents system information
type SystemInfo struct {
	OSName            string `json:"osName"`
	OSVersion         string `json:"osVersion"`
	KernelVersion     string `json:"kernelVersion"`
	Hostname          string `json:"hostname"`
	BootTime          uint64 `json:"bootTime"`
	UptimeSeconds     uint64 `json:"uptimeSeconds"`
	MotherboardModel  string `json:"motherboardModel,omitempty"`
	MotherboardVendor string `json:"motherboardVendor,omitempty"`
	BIOSVersion       string `json:"biosVersion,omitempty"`
}

// ProcessInfo represents process information
type ProcessInfo struct {
	PID         int     `json:"pid"`
	Name        string  `json:"name"`
	User        string  `json:"user"`
	CPUPercent  float64 `json:"cpuPercent"`
	MemoryBytes uint64  `json:"memoryBytes"`
	Status      string  `json:"status"`
	StartTime   int64   `json:"startTime"`
}

// ContainerInfo represents Docker container information
type ContainerInfo struct {
	ID      string `json:"id"`
	Name    string `json:"name"`
	Image   string `json:"image"`
	Status  string `json:"status"`
	State   string `json:"state"`
	Created int64  `json:"created"`
}

// MetricsType indicates the type of metrics message
type MetricsType int

const (
	MetricsFull     MetricsType = 0
	MetricsRealtime MetricsType = 1
	MetricsStatic   MetricsType = 2
	MetricsPeriodic MetricsType = 3
)

// DataRequestType indicates what data the server is requesting
type DataRequestType int

const (
	DataRequestFull     DataRequestType = 0
	DataRequestStatic   DataRequestType = 1
	DataRequestPeriodic DataRequestType = 2
)

// DiskIO represents disk I/O metrics for realtime data
type DiskIO struct {
	Device           string `json:"device"`
	ReadBytesPerSec  uint64 `json:"readBytesPerSec"`
	WriteBytesPerSec uint64 `json:"writeBytesPerSec"`
}

// NetworkIO represents network I/O metrics for realtime data
type NetworkIO struct {
	Interface     string `json:"interface"`
	RxBytesPerSec uint64 `json:"rxBytesPerSec"`
	TxBytesPerSec uint64 `json:"txBytesPerSec"`
}

// GPUUsage represents lightweight GPU usage for realtime data
type GPUUsage struct {
	Index        uint32  `json:"index"`
	UsagePercent float64 `json:"usagePercent"`
	MemoryUsed   uint64  `json:"memoryUsed"`
	Temperature  float64 `json:"temperature"`
}

// NPUUsage represents lightweight NPU usage for realtime data
type NPUUsage struct {
	Index        uint32  `json:"index"`
	UsagePercent float64 `json:"usagePercent"`
	MemoryUsed   uint64  `json:"memoryUsed"`
	Temperature  float64 `json:"temperature"`
}

// RealtimeMetrics represents high-frequency metrics (sent every ~1 second)
type RealtimeMetrics struct {
	Timestamp      int64       `json:"timestamp"`
	Hostname       string      `json:"hostname"`
	CPUUsage       float64     `json:"cpuUsage"`
	CPUPerCore     []float64   `json:"cpuPerCore,omitempty"`
	MemoryUsed     uint64      `json:"memoryUsed"`
	MemoryPercent  float64     `json:"memoryPercent"`
	SwapUsed       uint64      `json:"swapUsed"`
	DiskIO         []DiskIO    `json:"diskIo,omitempty"`
	NetworkIO      []NetworkIO `json:"networkIo,omitempty"`
	GPUUsages      []GPUUsage  `json:"gpuUsages,omitempty"`
	NPUUsages      []NPUUsage  `json:"npuUsages,omitempty"`
	LoadAverage    []float64   `json:"loadAverage,omitempty"`
	CPUTemperature float64     `json:"cpuTemperature,omitempty"`
}

// CPUStaticInfo represents static CPU information
type CPUStaticInfo struct {
	Model        string `json:"model"`
	Vendor       string `json:"vendor"`
	Cores        int    `json:"cores"`
	Threads      int    `json:"threads"`
	FrequencyMHz uint64 `json:"frequencyMhz"`
	CacheSize    uint64 `json:"cacheSize"`
	Architecture string `json:"architecture"`
}

// MemoryStaticInfo represents static memory information
type MemoryStaticInfo struct {
	TotalPhysical uint64 `json:"totalPhysical"`
	TotalSwap     uint64 `json:"totalSwap"`
	MemoryType    string `json:"memoryType,omitempty"`
	SpeedMHz      uint32 `json:"speedMhz,omitempty"`
	Slots         uint32 `json:"slots,omitempty"`
}

// DiskStaticInfo represents static disk information
type DiskStaticInfo struct {
	Device     string `json:"device"`
	Model      string `json:"model,omitempty"`
	Serial     string `json:"serial,omitempty"`
	Type       string `json:"type"` // SSD, HDD, NVMe
	Total      uint64 `json:"total"`
	FSType     string `json:"fsType"`
	MountPoint string `json:"mountPoint"`
}

// NetworkStaticInfo represents static network interface information
type NetworkStaticInfo struct {
	Interface  string   `json:"interface"`
	MacAddress string   `json:"macAddress"`
	SpeedMbps  uint64   `json:"speedMbps,omitempty"`
	Type       string   `json:"type,omitempty"` // ethernet, wifi, virtual
	IPAddress  []string `json:"ipAddress,omitempty"`
}

// GPUStaticInfo represents static GPU information
type GPUStaticInfo struct {
	Index          uint32 `json:"index"`
	Name           string `json:"name"`
	Vendor         string `json:"vendor"`
	MemoryTotal    uint64 `json:"memoryTotal"`
	DriverVersion  string `json:"driverVersion"`
	PCIeGeneration uint32 `json:"pcieGeneration,omitempty"`
}

// NPUStaticInfo represents static NPU information
type NPUStaticInfo struct {
	Index         uint32 `json:"index"`
	Name          string `json:"name"`
	Vendor        string `json:"vendor"`
	MemoryTotal   uint64 `json:"memoryTotal"`
	DriverVersion string `json:"driverVersion"`
}

// StaticInfo represents hardware information that rarely changes (sent once on connect)
type StaticInfo struct {
	Timestamp         int64               `json:"timestamp"`
	Hostname          string              `json:"hostname"`
	OSName            string              `json:"osName"`
	OSVersion         string              `json:"osVersion"`
	KernelVersion     string              `json:"kernelVersion"`
	BootTime          uint64              `json:"bootTime"`
	MotherboardModel  string              `json:"motherboardModel,omitempty"`
	MotherboardVendor string              `json:"motherboardVendor,omitempty"`
	BIOSVersion       string              `json:"biosVersion,omitempty"`
	CPU               *CPUStaticInfo      `json:"cpu,omitempty"`
	Memory            *MemoryStaticInfo   `json:"memory,omitempty"`
	Disks             []DiskStaticInfo    `json:"disks,omitempty"`
	Networks          []NetworkStaticInfo `json:"networks,omitempty"`
	GPUs              []GPUStaticInfo     `json:"gpus,omitempty"`
	NPUs              []NPUStaticInfo     `json:"npus,omitempty"`
}

// DiskUsage represents periodic disk usage update
type DiskUsage struct {
	MountPoint string `json:"mountPoint"`
	Used       uint64 `json:"used"`
	Available  uint64 `json:"available"`
}

// NetworkAddressUpdate represents periodic network address update
type NetworkAddressUpdate struct {
	Interface   string   `json:"interface"`
	IPAddresses []string `json:"ipAddresses"`
}

// PeriodicData represents data that changes slowly (sent every 30s-5min)
type PeriodicData struct {
	Timestamp      int64                  `json:"timestamp"`
	Hostname       string                 `json:"hostname"`
	UptimeSeconds  uint64                 `json:"uptimeSeconds"`
	DiskUsage      []DiskUsage            `json:"diskUsage,omitempty"`
	NetworkAddress []NetworkAddressUpdate `json:"networkAddress,omitempty"`
	UserSessions   []UserSession          `json:"userSessions,omitempty"`
}

File: sdk/python/nanolink/metrics.py
"""
Metrics data models for NanoLink SDK
"""

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class CpuMetrics:
    """CPU metrics data"""
    usage_percent: float = 0.0
    core_count: int = 0
    per_core_usage: List[float] = field(default_factory=list)
    model: str = ""
    vendor: str = ""
    frequency_mhz: float = 0.0
    max_frequency_mhz: float = 0.0
    temperature_celsius: Optional[float] = None
    architecture: str = ""


@dataclass
class MemoryMetrics:
    """Memory metrics data"""
    total: int = 0
    used: int = 0
    available: int = 0
    swap_total: int = 0
    swap_used: int = 0
    cached: int = 0
    buffers: int = 0
    memory_type: str = ""  # DDR4, DDR5, etc.
    speed_mhz: int = 0

    @property
    def usage_percent(self) -> float:
        """Calculate memory usage percentage"""
        if self.total == 0:
            return 0.0
        return (self.used / self.total) * 100

    @property
    def swap_usage_percent(self) -> float:
        """Calculate swap usage percentage"""
        if self.swap_total == 0:
            return 0.0
        return (self.swap_used / self.swap_total) * 100


@dataclass
class DiskMetrics:
    """Disk metrics data"""
    mount_point: str = ""
    device: str = ""
    fs_type: str = ""
    total: int = 0
    used: int = 0
    available: int = 0
    read_bytes_per_sec: int = 0
    write_bytes_per_sec: int = 0
    read_iops: int = 0
    write_iops: int = 0
    model: str = ""
    serial: str = ""
    disk_type: str = ""  # SSD, HDD, NVMe
    temperature_celsius: Optional[float] = None
    health_status: str = ""

    @property
    def usage_percent(self) -> float:
        """Calculate disk usage percentage"""
        if self.total == 0:
            return 0.0
        return (self.used / self.total) * 100


@dataclass
class NetworkMetrics:
    """Network interface metrics data"""
    interface: str = ""
    rx_bytes_per_sec: int = 0
    tx_bytes_per_sec: int = 0
    rx_packets_per_sec: int = 0
    tx_packets_per_sec: int = 0
    is_up: bool = False
    mac_address: str = ""
    ip_addresses: List[str] = field(default_factory=list)
    link_speed_mbps: int = 0
    interface_type: str = ""  # ethernet, wifi, loopback


@dataclass
class GpuMetrics:
    """GPU metrics data"""
    index: int = 0
    name: str = ""
    vendor: str = ""  # NVIDIA, AMD, Intel
    usage_percent: float = 0.0
    memory_total: int = 0
    memory_used: int = 0
    temperature_celsius: float = 0.0
    fan_speed_percent: float = 0.0
    power_draw_watts: float = 0.0
    power_limit_watts: float = 0.0
    clock_core_mhz: int = 0
    clock_memory_mhz: int = 0
    driver_version: str = ""
    pcie_gen: int = 0
    pcie_width: int = 0
    encoder_usage_percent: float = 0.0
    decoder_usage_percent: float = 0.0

    @property
    def memory_usage_percent(self) -> float:
        """Calculate GPU memory usage percentage"""
        if self.memory_total == 0:
            return 0.0
        return (self.memory_used / self.memory_total) * 100


@dataclass
class SystemInfo:
    """System information data"""
    os_name: str = ""
    os_version: str = ""
    kernel_version: str = ""
    hostname: str = ""
    boot_time: int = 0
    uptime_seconds: int = 0
    motherboard_model: str = ""
    motherboard_vendor: str = ""
    bios_version: str = ""


@dataclass
class Metrics:
    """Complete system metrics from an agent"""
    timestamp: int = 0
    hostname: str = ""
    cpu: Optional[CpuMetrics] = None
    memory: Optional[MemoryMetrics] = None
    disks: List[DiskMetrics] = field(default_factory=list)
    networks: List[NetworkMetrics] = field(default_factory=list)
    gpus: List[GpuMetrics] = field(default_factory=list)
    system: Optional[SystemInfo] = None
    load_average: List[float] = field(default_factory=list)

    @classmethod
    def from_dict(cls, data: dict) -> "Metrics":
        """Create Metrics from dictionary"""
        metrics = cls()
        metrics.timestamp = data.get("timestamp", 0)
        metrics.hostname = data.get("hostname", "")

        if "cpu" in data and data["cpu"]:
            cpu_data = data["cpu"]
            metrics.cpu = CpuMetrics(
                usage_percent=cpu_data.get("usagePercent", 0.0),
                core_count=cpu_data.get("coreCount", 0),
                per_core_usage=cpu_data.get("perCoreUsage", []),
                model=cpu_data.get("model", ""),
                vendor=cpu_data.get("vendor", ""),
                frequency_mhz=cpu_data.get("frequencyMhz", 0.0),
                max_frequency_mhz=cpu_data.get("maxFrequencyMhz", 0.0),
                temperature_celsius=cpu_data.get("temperatureCelsius"),
                architecture=cpu_data.get("architecture", ""),
            )

        if "memory" in data and data["memory"]:
            mem_data = data["memory"]
            metrics.memory = MemoryMetrics(
                total=mem_data.get("total", 0),
                used=mem_data.get("used", 0),
                available=mem_data.get("available", 0),
                swap_total=mem_data.get("swapTotal", 0),
                swap_used=mem_data.get("swapUsed", 0),
                cached=mem_data.get("cached", 0),
                buffers=mem_data.get("buffers", 0),
                memory_type=mem_data.get("memoryType", ""),
                speed_mhz=mem_data.get("speedMhz", 0),
            )

        if "disks" in data:
            for disk_data in data["disks"]:
                metrics.disks.append(DiskMetrics(
                    mount_point=disk_data.get("mountPoint", ""),
                    device=disk_data.get("device", ""),
                    fs_type=disk_data.get("fsType", ""),
                    total=disk_data.get("total", 0),
                    used=disk_data.get("used", 0),
                    available=disk_data.get("available", 0),
                    read_bytes_per_sec=disk_data.get("readBytesPerSec", 0),
                    write_bytes_per_sec=disk_data.get("writeBytesPerSec", 0),
                    read_iops=disk_data.get("readIops", 0),
                    write_iops=disk_data.get("writeIops", 0),
                    model=disk_data.get("model", ""),
                    serial=disk_data.get("serial", ""),
                    disk_type=disk_data.get("diskType", ""),
                    temperature_celsius=disk_data.get("temperatureCelsius"),
                    health_status=disk_data.get("healthStatus", ""),
                ))

        if "networks" in data:
            for net_data in data["networks"]:
                metrics.networks.append(NetworkMetrics(
                    interface=net_data.get("interface", ""),
                    rx_bytes_per_sec=net_data.get("rxBytesPerSec", 0),
                    tx_bytes_per_sec=net_data.get("txBytesPerSec", 0),
                    rx_packets_per_sec=net_data.get("rxPacketsPerSec", 0),
                    tx_packets_per_sec=net_data.get("txPacketsPerSec", 0),
                    is_up=net_data.get("isUp", False),
                    mac_address=net_data.get("macAddress", ""),
                    ip_addresses=net_data.get("ipAddresses", []),
                    link_speed_mbps=net_data.get("linkSpeedMbps", 0),
                    interface_type=net_data.get("interfaceType", ""),
                ))

        if "gpus" in data:
            for gpu_data in data["gpus"]:
                metrics.gpus.append(GpuMetrics(
                    index=gpu_data.get("index", 0),
                    name=gpu_data.get("name", ""),
                    vendor=gpu_data.get("vendor", ""),
                    usage_percent=gpu_data.get("usagePercent", 0.0),
                    memory_total=gpu_data.get("memoryTotal", 0),
                    memory_used=gpu_data.get("memoryUsed", 0),
                    temperature_celsius=gpu_data.get("temperatureCelsius", 0.0),
                    fan_speed_percent=gpu_data.get("fanSpeedPercent", 0.0),
                    power_draw_watts=gpu_data.get("powerDrawWatts", 0.0),
                    power_limit_watts=gpu_data.get("powerLimitWatts", 0.0),
                    clock_core_mhz=gpu_data.get("clockCoreMhz", 0),
                    clock_memory_mhz=gpu_data.get("clockMemoryMhz", 0),
                    driver_version=gpu_data.get("driverVersion", ""),
                    pcie_gen=gpu_data.get("pcieGen", 0),
                    pcie_width=gpu_data.get("pcieWidth", 0),
                    encoder_usage_percent=gpu_data.get("encoderUsagePercent", 0.0),
                    decoder_usage_percent=gpu_data.get("decoderUsagePercent", 0.0),
                ))

        if "system" in data and data["system"]:
            sys_data = data["system"]
            metrics.system = SystemInfo(
                os_name=sys_data.get("osName", ""),
                os_version=sys_data.get("osVersion", ""),
                kernel_version=sys_data.get("kernelVersion", ""),
                hostname=sys_data.get("hostname", ""),
                boot_time=sys_data.get("bootTime", 0),
                uptime_seconds=sys_data.get("uptimeSeconds", 0),
                motherboard_model=sys_data.get("motherboardModel", ""),
                motherboard_vendor=sys_data.get("motherboardVendor", ""),
                bios_version=sys_data.get("biosVersion", ""),
            )

        metrics.load_average = data.get("loadAverage", [])

        return metrics
